cmd/crank/beta/trace/internal/printer/default.go

```diff
@@ -0,0 +1,192 @@
+			// prefix for its children
+		case item.isLast:
+			name.WriteString(item.prefix + "└─ ")
+			childPrefix += "   "
+		default:
+			name.WriteString(item.prefix + "├─ ")
+			childPrefix += "│  "
+		}
+
+		name.WriteString(fmt.Sprintf("%s/%s", item.resource.Unstructured.GetKind(), item.resource.Unstructured.GetName()))
+
+		// Append the namespace if it's not empty
+		if item.resource.Unstructured.GetNamespace() != "" {
+			name.WriteString(fmt.Sprintf(" (%s)", item.resource.Unstructured.GetNamespace()))
+		}
+		ready, synced, status := getResourceStatus(item.resource, p.wide)
+
+		row := defaultPrinterRow{
+			name:   name.String(),
+			ready:  ready,
+			synced: synced,
+			status: status,
+		}
+		if _, err := fmt.Fprintln(tw, row.String()); err != nil {
+			return errors.Wrap(err, errWriteRow)
+		}
+
+		// Enqueue the children of the current node in reverse order to ensure
+		// that they are dequeued from the LIFO queue in the same order w.r.t.
+		// the way they are defined by the resources.
+		for idx := len(item.resource.Children) - 1; idx >= 0; idx-- {
+			isLast := idx == len(item.resource.Children)-1
+			queue = append(queue, &queueItem{resource: item.resource.Children[idx], depth: item.depth + 1, isLast: isLast, prefix: childPrefix})
+		}
+	}
+
+	if err := tw.Flush(); err != nil {
+		return errors.Wrap(err, errFlushTabWriter)
+	}
+
+	return nil
+}
+
+// getResourceStatus returns the status of the resource.
+func getResourceStatus(r *resource.Resource, wide bool) (ready string, synced string, status string) {
+	readyCond := r.GetCondition(xpv1.TypeReady)
+	syncedCond := r.GetCondition(xpv1.TypeSynced)
+	var m string
+	switch {
+	case r.Error != nil:
+		// if there is an error we want to show it
+		status = "Error"
+		m = r.Error.Error()
+	case readyCond.Status == corev1.ConditionTrue && syncedCond.Status == corev1.ConditionTrue:
+		// if both are true we want to show the ready reason only
+		status = string(readyCond.Reason)
+
+	// The following cases are for when one of the conditions is false,
+	// prioritizing readiness over synced in case of issues.
+	case readyCond.Status == corev1.ConditionFalse:
+		status = string(readyCond.Reason)
+		m = readyCond.Message
+	case syncedCond.Status == corev1.ConditionFalse:
+		status = string(syncedCond.Reason)
+		m = syncedCond.Message
+
+	default:
+		// both are unknown or unset, let's try showing the ready reason, probably empty
+		status = string(readyCond.Reason)
+		m = readyCond.Message
+	}
+
+	// Crop the message to the last 64 characters if it's too long and we are
+	// not in wide mode
+	if !wide && len(m) > 64 {
+		m = "..." + m[len(m)-64:]
+	}
+
+	// Append the message to the status if it's not empty
+	if m != "" {
+		status = fmt.Sprintf("%s: %s", status, m)
+	}
+
+	return mapUnknownStatusToDash(readyCond.Status), mapUnknownStatusToDash(syncedCond.Status), status
+}
+
+func mapUnknownStatusToDash(s corev1.ConditionStatus) string {
+	if s == corev1.ConditionUnknown {
```

Comment:
I was under the impression that there are really cases where a resources `Synced` or `Ready` condition was set to `Unknown` and there are also cases where those condition not existing at all. We will still show the former as `Unknown` but the latter as `-`. However, it looks like we will be mapping both cases to `Unknown` with this implementation.

Reply:
the defaulting was happening at a lower level, on the ConditionedStatus.GetCondition, so I thought it would be fine and more compact to just always show "-", but you are right it make more sense not to throw away this information.
I've duplicated the GetCondition code and now return a really empty condition, so now we have:
- True/False/Unknown => True/False/Unknown
- "" => "-"

Reply:
https://github.com/crossplane/crossplane/pull/4849/commits/f8998ceee913e53c1c4585e448f5d0936c44fb38